base_layer/wallet/src/output_manager_service/storage/database.rs

```diff
@@ -0,0 +1,376 @@
+macro_rules! fetch {
+    ($self:ident, $key_val:expr, $key_var:ident) => {{
+        let key = DbKey::$key_var($key_val);
+        match $self.db.fetch(&key) {
+            Ok(None) => Err(OutputManagerStorageError::ValueNotFound(key)),
+            Ok(Some(DbValue::$key_var(k))) => Ok(*k),
+            Ok(Some(other)) => unexpected_result(key, other),
+            Err(e) => log_error(key, e),
+        }
+    }};
+}
+
+/// This structure holds an inner type that implements the `OutputManagerBackend` trait and contains the more complex
+/// data access logic required by the module built onto the functionality defined by the trait
+pub struct OutputManagerDatabase<T>
+where T: OutputManagerBackend
+{
+    db: T,
+}
+
+impl<T> OutputManagerDatabase<T>
+where T: OutputManagerBackend
+{
+    pub fn new(db: T) -> Self {
+        Self { db }
+    }
+
+    pub fn add_unspent_output(&mut self, output: UnblindedOutput) -> Result<(), OutputManagerStorageError> {
+        self.db.write(WriteOperation::Insert(DbKeyValuePair::UnspentOutput(
+            output.spending_key.clone(),
+            Box::new(output),
+        )))?;
+
+        Ok(())
+    }
+
+    pub fn get_balance(&self) -> Result<MicroTari, OutputManagerStorageError> {
+        if let DbValue::UnspentOutputs(uo) =
+            self.db
+                .fetch(&DbKey::UnspentOutputs)?
+                .ok_or(OutputManagerStorageError::UnexpectedResult(
+                    "Unspent Outputs cannot be retrieved".to_string(),
+                ))?
+        {
+            Ok(uo.iter().fold(MicroTari::from(0), |acc, x| acc + x.value))
+        } else {
+            Err(OutputManagerStorageError::UnexpectedResult(
+                "Unexpected result from database backend".to_string(),
+            ))
+        }
+    }
+
+    pub fn add_pending_transaction_outputs(
+        &mut self,
+        pending_transaction_ouputs: PendingTransactionOutputs,
+    ) -> Result<(), OutputManagerStorageError>
+    {
+        self.db
+            .write(WriteOperation::Insert(DbKeyValuePair::PendingTransactionOutputs(
+                pending_transaction_ouputs.tx_id.clone(),
+                Box::new(pending_transaction_ouputs),
+            )))?;
+
+        Ok(())
+    }
+
+    pub fn fetch_pending_transaction_outputs(
+        &self,
+        tx_id: TxId,
+    ) -> Result<PendingTransactionOutputs, OutputManagerStorageError>
+    {
+        fetch!(self, tx_id, PendingTransactionOutputs)
+    }
+
+    /// This method is called when a pending transaction is confirmed. It moves the `outputs_to_be_spent` and
+    /// `outputs_to_be_received` from a `PendingTransactionOutputs` record into the `unspent_outputs` and
+    /// `spent_outputs` collections.
+    pub fn confirm_pending_transaction_outputs(&mut self, tx_id: TxId) -> Result<(), OutputManagerStorageError> {
+        let pending_tx = fetch!(self, tx_id.clone(), PendingTransactionOutputs)?;
```

Comment:
It would be less opinionated and give backend more freedom in how to implement this, if there was a DbKey `ConfirmTx(tx_id)` and you left the backend to figure out what that means.